COV-144, Add service for determine covid status based on ct values

GitEdvard opened the pull request:
Add code to determine covid19 status based on ct values and assay. If one of the two rtPCR controls fails, the entire plate is failed. 

Points for careful attention
* the udf "Has Covid19" is a text field with 3 valid values, "Yes", "No" and "Undefined"
* A plate may be passed, but individual samples may still have undefined results
* The boundaries to validate controls are the same as deciding on sample results. 
* This setup assumes that only one plate is entered into the rtPCR step at a time. 

This code is not tested yet. My plan is to get help from Ingvar to setup the step and udfs, and then test it.

johandahlberg commented on clarity-ext-scripts/clarity_ext_scripts/covid/ct_discriminator.py:
```diff
@@ -0,0 +1,66 @@
+
+    def _has_assay_udf(self):
+        try:
+            dummy = self.assay
```
There should be no need to assign a dummy variable to get the desired result here.

withrocks replied:
I think some code quality tools might complain if it's only `self.assay`, so perhaps `_ = self.assay`? Of course `hasattr` would work generally, but not entirely sure in this case, as this is our own UdfMapping stuff, not sure it supports that, because they use the black magic `get_attribute` family of methods.

johandahlberg commented on clarity-ext-scripts/clarity_ext_scripts/covid/ct_discriminator.py:
```diff
@@ -0,0 +1,66 @@
+
+    def _get_discriminator_values(self):
+        """
+        Returns: lower bound, upper bound
+        ct <= lower bound: counted as a positive value
+        ct >=upper bound: is counted as a negative value
+        lower bound < ct < upper bound: undefined
```
I'm not sure that the bounds, as defined here, is actually how things will work here. Reading through the details of this, I'm not even sure that having the Ct value is enough to make the call about the diagnosis.

From the manual Maike provided:

**Quality control**
 - Blank control: Ct values at FAM and VIC/HEX channels are 0 or no data available. 
 - Positive control：Standard curves at channel FAM and VIC/HEX channels are in S-shape with Ct values not higher than 32. 
 - Testing specimen: Standard curves at VIC/HEX channel is in S-shape with Ct not higher than 32. 
 - Above requirements should be met in a single test. Otherwise, the test is invalid. Please operate the retest strictly in line with the package insert. 

**Threshold and reference**
 - rangeCut-off value of the kit was determined based on the Receiver Operator characteristic curve from testing clinical samples. Ct value for 2019-nCoV positive by the kit is not high than 38.

**Testing result interpretation**
 - The specimen is positive of 2019-nCoV if standard curve at FAM channel is in S-shape with Ct value not higher than 38. 
 - The specimen is negative of 2019-nCoV if standard curve at FAM channel is not in S-shape with Ct at FAM as 0 or no data available while Ct at VIC/HEX not higher than 32.
 - The specimen should be retested if standard curve at FAM is in S-shape with Ct higher than 38. The specimen can be reported on basis of retesting results as positive of 2019-nCoV for Ct higher than 38 and as negative of 2019-nCoV for standard curve not in S-shape and Ct of internal reference not higher than 32 at VIC/HEX. 
 - In case that standard curve at FAM is not in S-shape with Ct value as 0 or no data available, the specimen should be retested if Ct at VIC is higher than 32 or no data available.

withrocks replied:
Maybe @kleppari wants to join this discussion?

johandahlberg commented on clarity-ext-scripts/clarity_ext_scripts/covid/ct_discriminator.py:
```diff
@@ -0,0 +1,66 @@
+
+    def analyze(self, ct_value):
+        """
+        Returns: String with value of either 'Yes', 'No', 'Undefined'
```
It think these should rather correspond to the values that we want to use in the reporting later, i.e. positive/negative/failed, unless there is some other Clarity related reason for using these values instead.

withrocks replied:
I agree with that. We should be able to use any string here (we're in control of the validation in Clarity)

GitEdvard replied:
ok

johandahlberg commented on clarity-ext-scripts/clarity_ext_scripts/covid/ct_analyes_service.py:
```diff
@@ -0,0 +1,37 @@
+            original_sample = output.sample()
+            ct = output.udf_ct
+            has_covid = ct_analysis_service.analyze(ct)
+            original_sample.udf_map.force("Has Covid19", has_covid)
```
I think that the naming scheme here should rather be something a long the lines of "qcPCR covid-19 result", which also works better with my suggestion below about having values positive/negative/failed.

withrocks replied:
Right. On top of that, correct me if I'm wrong, but the test doesn't really say anything about the disease (covid19) but only the virus (SARS-CoV-2). But I guess this will be understood correctly in any case.

GitEdvard replied:
ok

johandahlberg commented on clarity-ext-scripts/clarity_ext_scripts/covid/ct_analyes_service.py:
```diff
@@ -0,0 +1,37 @@
+from clarity_ext.domain.validation import UsageError
+
+
+class CtAnalyseService(object):
```
Maybe `QPCRAnalysisService` is a better name here, since we might actually need to take more than the Ct values into account. See my extract from the manual below.

withrocks replied:
Right, I think it might be easier to understand from a dev's point of view too.

GitEdvard replied:
ok

GitEdvard replied:
I'll use RTPCR if you don't mind

withrocks commented on clarity-ext-scripts/clarity_ext_scripts/covid/parse_pcr.py:
```diff
@@ -40,6 +43,7 @@ def execute(self):
             original_sample.udf_map.force("CT source", output.api_resource.uri)
             self.context.update(original_sample)
 
+        ct_analyes_service.execute()
```
Is there a typo here? s/ct_analyes/ct_analytes ?

GitEdvard replied:
yes

GitEdvard commented:
Now, I added two more commits addressing the comments above. Here is my proposal to the testability issue:

CtDiscriminator is free from Clarity code, RTPCRAnalyseService is not. If more things is to be tested, then we move code to the CtDiscriminator. There is a need for a class with Clarity code, populating all udfs. 

The CtDiscriminator code is rather thin as it is now. It will be changed though.

johandahlberg commented:
Thank you. My suggestion for a separation here is:

The extensions scripts have the connection to Clarity, while the services are feed only the information they need. Does that make sense to you?
